### ISSUE TITLE
std::sort/swap compiler error starting with the VS2022 17.7.0 update

### ISSUE THREAD

**Issue by foundry-markf** [microsoft/STL#3966]
# Describe the bug
This error below is from thirdparty code from Qt 5 that we compile for Windows, Linux and Mac. This code hasn't been changed for over a year. 

It only started showing up when upgrading to VisualStudio 2022 17.7.0 recently. Specifically I see it on this version

`Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32822 for x64`

I wasn't sure whether to ask a question or raise a bug, but since it only fails to compile in one specific version of VS2022, and I have a reproducible, I thought it best to go for bug.
The error that now appears starts with (I'll give the full error messge below)

`C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): error C2672: 'swap': no matching overloaded function found`

The same code compiles fine with the following compiler versions

`Microsoft (R) C/C++ Optimizing Compiler Version 19.36.32535 for x64`
`Microsoft (R) C/C++ Optimizing Compiler Version 19.35.32216.1 for x64`

and likely several more before that (including VS2019) as we've been using both compilers on this source tree for a while.

We use GCC 11 on Linux and Apple Clang 13 on Mac. Both compile the code without error.

# Command-line test case
I took the Qt code (it was Chromium to be fair) and simplified it down to the minimum I could find to reproduce it. It is within a thirdparty library called Perfetto.

I have put it into Compiler Explorer, which you can see here: https://godbolt.org/z/EeYzzxacr This is a verbatim copy-n-past of the code I'm testing in-house as a reproducible. I'm happy to provide this by other means too if necessary (e.g. a gist).

The link tests against GCC11, GCC13, Clang 16, and MSVC Latest, which happens to be older than the version I have from the 17.7.0 update:

`Microsoft (R) C/C++ Optimizing Compiler Version 19.36.32530 for x64`

so it looks like Compiler Explorer do not have the bleeding edge.

But, it shows all four compilers compiling successfully. 

Locally, I am compiling that source code with
```
cl.exe -std:c++14 -c trace_sorter.cc
```
to reproduce the error.

# Expected behavior
I was surprised of the change of behaviour with the 17.7.0 update.

We've discussed the error internally, and think the problem is to do with constness, but uncertain whether it's a change in the STL implementation or very old thirdparty code that wasn't ever compliant and just lucked out. As much as other implementations have their quirks, VS2022 did used to compile this, and the other compilers are still compiling this code successfully.

I did notice this change recently, https://github.com/microsoft/STL/commit/71e404b2974ef0045851fc79da16cce6edc0aaf1, but am not convinced if this is related.

I am curious if anyone can pinpoint the exact change that causes the behaviour change, so that we can decide what we need to do to make this work again.

# STL version

Microsoft Visual Studio Professional 2022
Version 17.7.0

# Additional context
Full error message
```
$ . 'C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\Tools\Launch-VsDevShell.ps1' -Arch amd64 -SkipAutomaticLocation
**********************************************************************
** Visual Studio 2022 Developer PowerShell v17.7.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
$ cl.exe -std:c++14 -c trace_sorter.cc
Microsoft (R) C/C++ Optimizing Compiler Version 19.37.32822 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
trace_sorter.cc
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): error C2672: 'swap': no matching overloaded function found
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\utility(468): note: could be 'void std::swap(std::pair<_Ty1,_Ty2> &,std::pair<_Ty1,_Ty2> &) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): note: 'void std::swap(std::pair<_Ty1,_Ty2> &,std::pair<_Ty1,_Ty2> &) noexcept(<expr>)': could not deduce template argument for 'std::pair<_Ty1,_Ty2> &' from 'T'
        with
        [
            T=IntWrapper
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\type_traits([21](https://gitlab.thefoundry.co.uk/mark.final/vs2022_177_regression_test/-/jobs/660207#L21)56): note: or       'void std::swap(_Ty (&)[_Size],_Ty (&)[_Size]) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\20[22](https://gitlab.thefoundry.co.uk/mark.final/vs2022_177_regression_test/-/jobs/660207#L22)\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): note: 'void std::swap(_Ty (&)[_Size],_Ty (&)[_Size]) noexcept(<expr>)': could not deduce template argument for '_Ty (&)[_Size]' from 'T'
        with
        [
            T=IntWrapper
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\type_traits(2153): note: or       'void std::swap(_Ty &,_Ty &) noexcept(<expr>)'
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): note: 'void std::swap(_Ty &,_Ty &) noexcept(<expr>)': template parameter '_Ty' is ambiguous
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): note: could be 'const T'
        with
        [
            T=IntWrapper
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): note: or       'T'
        with
        [
            T=IntWrapper
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8002): note: 'void std::swap(_Ty &,_Ty &) noexcept(<expr>)': could not deduce template argument for '_Ty &' from 'const T'
        with
        [
            T=IntWrapper
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8048): note: see reference to function template instantiation 'std::pair<_RanIt,_RanIt> std::_Partition_by_median_guess_unchecked<_RanIt,_Pr>(_RanIt,_RanIt,_Pr)' being compiled
        with
        [
            _RanIt=CircularQueue<IntWrapper>::Iterator,
            _Pr=std::less<void>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8067): note: see reference to function template instantiation 'void std::_Sort_unchecked<_RanIt,_Fn>(_RanIt,_RanIt,__int64,_Pr)' being compiled
        with
        [
            _RanIt=CircularQueue<IntWrapper>::Iterator,
            _Fn=std::less<void>,
            _Pr=std::less<void>
        ]
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.37.32822\include\algorithm(8072): note: see reference to function template instantiation 'void std::sort<_RanIt,std::less<void>>(const _RanIt,const _RanIt,_Pr)' being compiled
        with
        [
            _RanIt=CircularQueue<IntWrapper>::Iterator,
            _Pr=std::less<void>
        ]
trace_sorter.cc(114): note: see reference to function template instantiation 'void std::sort<CircularQueue<IntWrapper>::Iterator>(const _RanIt,const _RanIt)' being compiled
        with
        [
            _RanIt=CircularQueue<IntWrapper>::Iterator
        ]
```

**Comment by achabense** [comment]
<details>
<summary>Smaller repro</summary>

```cpp
#include <algorithm>
#include <cstddef>

struct IntWrapper {
    inline bool operator<(const IntWrapper& o) const {
        return value < o.value;
    }

    IntWrapper& operator=(const IntWrapper& o) const {
        IntWrapper& non_const_this = const_cast<IntWrapper&>(*this);
        non_const_this.value = o.value;
        return non_const_this;
    }

    int value;
};


template<class T>
class Iterator {
public:
    using difference_type = ptrdiff_t;
    using value_type = T;
    using pointer = const T*;
    using reference = const T&;
    using iterator_category = std::random_access_iterator_tag;

    Iterator(int* queue, unsigned int pos, unsigned int generation)
        : queue_(queue),
        pos_(pos)
    {
    }

    T* operator->() {
        return nullptr;
    }

    const T* operator->() const {
        return nullptr;
    }

    T& operator*() { return *(operator->()); }
    const T& operator*() const { return *(operator->()); }

    Iterator& operator++() {
        return *this;
    }

    Iterator& operator--() {
        return *this;
    }

    friend Iterator operator+(const Iterator& iter, difference_type offset) {
        return iter;
    }

    Iterator& operator+=(difference_type offset) {
        return *this;
    }

    friend Iterator operator-(const Iterator& iter, difference_type offset) {
        return iter;
    }

    friend ptrdiff_t operator-(const Iterator& lhs, const Iterator& rhs) {
        return static_cast<ptrdiff_t>(lhs.pos_) -
            static_cast<ptrdiff_t>(rhs.pos_);
    }

    friend bool operator==(const Iterator& lhs, const Iterator& rhs) {
        return lhs.pos_ == rhs.pos_;
    }

    friend bool operator!=(const Iterator& lhs, const Iterator& rhs) {
        return lhs.pos_ != rhs.pos_;
    }

    friend bool operator<(const Iterator& lhs, const Iterator& rhs) {
        return lhs.pos_ < rhs.pos_;
    }

private:
    int* queue_;
    unsigned int pos_;
};

namespace std {
    void repro() {
        using Iter = ::Iterator<IntWrapper>;
        auto x = _Get_unwrapped(Iter(nullptr, 0, 0));
        auto a = Iter(nullptr, 0U, 0U);
        const auto b = _Prev_iter(a); // mimicking `const auto _Glast_prev = _Prev_iter(_Glast);`
        swap(*a, *b);
    }
}
int main() {
    std::repro();
}
```
</details>

What happens to the swap:
![image](https://github.com/microsoft/STL/assets/60953653/4d6a12c1-db10-4d7c-a63c-b87495a23ff3)

**Comment by achabense** [comment]
```cpp
    T& operator*() { return *(operator->()); }
    const T& operator*() const { return *(operator->()); }
```
This looks not correct for me.

**Comment by foundry-markf** [comment]
For further reference, this is where the Iterator originates from

https://github.com/qt/qtwebengine-chromium/blob/v5.15.2/chromium/third_party/perfetto/include/perfetto/ext/base/circular_queue.h#L84-L85

**Comment by CaseyCarter** [comment]
> thirdparty code that wasn't ever compliant and just lucked out

This. `std::sort` requires the type of its arguments `first` and `last` to meet the _`Cpp17RandomAccessIterator`_ and _`Cpp17ValueSwappable`_ requirements. `CircularQueue::Iterator` doesn't meet several requirements:
1. `Iterator` is not default-constructible. It must be default-constructible, and value-initialized iterators must compare equal (`Iterator{} == Iterator{}` must be well-defined and always `true`).
2. `pointer` must be `decltype(a.operator->())` and `reference` must be `decltype(a.operator*())` "for an iterator object `a` of class type" ([iterator.traits/1). The _`Cpp17InputIterator`_ requirements (which are incorporated by the _`Cpp17RandomAccessIterator`_ requirements) require the expressions `a->m` and `*a` to be valid when `a` is a value of type `Iterator` or `const Iterator`. Together, these implicitly require that `Iterator::operator->()` and `Iterator::operator->() const` (respectively `Iterator::operator*()` and `Iterator::operator*() const`) have the same return type `pointer` (respectively `reference`). TLDR: `const`-ness should not affect iterators exactly has it doesn't effect pointers; `*p` and `p->meow` don't care if `p` is `const`.
3. `a[n]` must be "convertible to `reference`" (ditto "this shouldn't depend on `const`") ([tab:randomaccessiterator]).
4. "iterator + integer" works, but `Iterator` lacks symmetric support for "integer + iterator" ([tab:randomaccessiterator]).

I _think_ that's all of the issues, but I didn't exactly audit it thoroughly.

**Comment by foundry-markf** [comment]
Thank you for the code audit.

I took another look at the vanilla thirdparty code recently to look for hints as to whether it had been fixed upstream, armed with the knowledge in this issue. From the look of the history, they have struggled compiling with VisualStudio for a while

https://github.com/google/perfetto/commits/master/include/perfetto/ext/base/circular_queue.h

In particular, this commit, https://github.com/google/perfetto/commit/b82a213bcf71dd8e68b53941375936c7905c0ff2, is what I have (partially) taken into a fork.

I'm adding this detail for anyone who encounters the issue going forward, but also in the off chance that the Microsoft STL team have lists of external libraries to test against, or reach out to for those that have struggled to compile.
